doc/source/raysgd/v2/user_guide.rst

```diff
@@ -0,0 +1,308 @@
+        config={"worker_batch_size": 64},
+        dataset=dataset)
+
+
+.. note:: This feature currently does not work with elastic training.
+
+
+Monitoring training
+-------------------
+
+You may want to plug in your training code with your favorite experiment management framework.
+RaySGD provides an interface to fetch intermediate results and callbacks to process/log your intermediate results.
+
+You can plug all of these into RaySGD with the following interface:
+
+.. code-block:: python
+
+    def train_func(config):
+        # do something
+        for x, y in dataset:
+            result = process(x)
+            ray.sgd.report(**result)
+
+
+    # TODO: Where do we pass in the logging folder?
+    result = trainer.run(
+        train_func,
+        config={"worker_batch_size": 64},
+        callbacks=[sgd.MlflowCallback()]
+        dataset=dataset)
+
+.. Here is a list of callbacks that is supported by RaySGD:
+
+.. * WandbCallback
+.. * MlflowCallback
+.. * TensorboardCallback
+.. * JsonCallback (Automatically logs given parameters)
+.. * CSVCallback
+
+
+.. note:: When using RayTune, these callbacks will not be used.
+
+Checkpointing
+-------------
+
+RaySGD provides a way to save state during the training process. This will be useful for:
+
+1. :ref:`Integration with Ray Tune <tune-sgd>` to use certain Ray Tune schedulers
+2. Running a long-running training job on a cluster of pre-emptible machines/pods.
+
+
+.. code-block:: python
+
+    import ray
+
+    def train_func(*args):
+
+        state = ray.sgd.load_checkpoint()
+        # eventually, optional:
+        for _ in config["num_epochs"]:
+            train(...)
+            ray.sgd.save_checkpoint((model, optimizer, etc))
+        return model
+
+    trainer = Trainer(gpus_per_worker, cpus_per_worker, num_workers=(4, inf))
+    trainer.run(train_func)
+    state = trainer.get_last_checkpoint()
+
+.. Running on the cloud
+.. --------------------
+
+.. Use RaySGD with the Ray cluster launcher by changing the following:
+
+.. .. code-block:: bash
+
+..     ray up cluster.yaml
+
+.. TODO.
+
+
+
+.. Running on pre-emptible machines
+.. --------------------------------
+
+.. You may want to
+
+.. TODO.
+
+
+.. _tune-sgd:
+
+Hyperparameter tuning
+---------------------
+
+Hyperparameter tuning with Ray Tune is natively supported with RaySGD. Specifically, you can take an existing training function and follow these steps:
+
+1. Call ``trainer.to_tune_trainable``, which will produce an object ("Trainable") that will be passed to Ray Tune.
+2. Call ``tune.run(trainable)`` instead of ``trainer.run``. This will invoke the hyperparameter tuning, starting multiple "trials" each with the resource amount specified by the Trainer.
+
+A couple caveats:
+
+* Tune won't handle the ``training_func`` return value correctly. To save your best trained model, you'll need to use the checkpointing API.
+* You should **not** call ``tune.report`` or ``tune.checkpoint_dir`` in your training function.
```

Comment:
We should add (maybe in a follow-up PR) the actual instructions on what interface to use (e.g. `sgd.report`) and what to expect this should translate to (`tune.report` on the result from one worker.)

Reply:
Yep agreed.